[WIP]  Add quantum number preserving gate fabric

obliviateandsurrender opened the pull request:
**Context:**
This PR adds a template for a local, expressive, and quantum number preserving ansatz based on [arXiv:2104.05695](https://arxiv.org/abs/2104.05695). 

**Description of the Change:**
Function `GateFabric` provides the implementation for the new template.
Function `OrbitalRotation` provides the implementation for the spatial orbital rotation gate used in generating `Q` gates for the gate fabric.
Relevant UnitTests for both  `GateFabric` and `OrbitalRotation`.

**Benefits:**
These functions extend the currently supported templates for the heuristics ansatze for fermionic systems with the VQE gate fabric with advantageous properties for the simulation of strongly correlated ground and excited states of molecules and materials under the Jordan-Wigner mapping as shown in [arXiv:2104.05695](https://arxiv.org/abs/2104.05695).

**Possible Drawbacks:**
No specific drawbacks.

**Related GitHub Issues:**
No related issues.

josh146 commented:
Thanks @obliviateandsurrender, this is a great new feature! Let us know if you have any questions (feel free to comment on particular lines of code). Otherwise, simply leave a comment here when this PR is ready for review :)

soranjh commented on pennylane/devices/autograd_ops.py:
```diff
@@ -386,3 +386,37 @@ def DoubleExcitationMinus(phi):
     return np.array(U)
+
+
+def QuantumNumberPreservingOR(theta):
```
The name of the parameter for this gate is different than the one used in the figure.

soranjh replied:
Also you might consider using `OrbitalRotation` for the name of the gates, instead of `QuantumNumberPreservingOR`. (Similarly,  `PairExchange` could be used, if we are implementing it.)

obliviateandsurrender replied:
1. I have changed the name of the gate to `OrbitalRotation`.
2. I have regenerated all the figures to match the parameters.

soranjh commented on pennylane/devices/autograd_ops.py:
```diff
@@ -386,3 +386,37 @@ def DoubleExcitationMinus(phi):
+
+    return np.array(
+        [
+            [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
```
Some of the signs are different than those in the paper, I recall that you found a typo in the original paper? If it is the case then we should document it here.

obliviateandsurrender replied:
I have documented the same now, along with the mention of the typo.

soranjh commented on pennylane/devices/jax_ops.py:
```diff
@@ -384,3 +384,38 @@ def DoubleExcitationMinus(phi):
     return jnp.array(U)
+
+
+def QuantumNumberPreservingOR(theta):
```
\phi?

obliviateandsurrender replied:
Fixed all the instances of parameter name mismatch.

soranjh commented on pennylane/devices/jax_ops.py:
```diff
@@ -384,3 +384,38 @@ def DoubleExcitationMinus(phi):
+    s = jnp.sin(theta / 2)
+
+    U = [
+        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
```
Some of the signs are different than those in the paper, please see my comment above.

obliviateandsurrender replied:
Done.

soranjh commented on pennylane/devices/tf_ops.py:
```diff
@@ -405,3 +405,38 @@ def DoubleExcitationMinus(phi):
     return tf.convert_to_tensor(U)
+
+
+def QuantumNumberPreservingOR(theta):
```
\phi?

soranjh commented on pennylane/devices/tf_ops.py:
```diff
@@ -405,3 +405,38 @@ def DoubleExcitationMinus(phi):
+    s = tf.sin(theta / 2)
+
+    U = [
+        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
```
Some of the signs are different than those in the paper, please see my comment above.

obliviateandsurrender replied:
Done.

soranjh commented on pennylane/devices/torch_ops.py:
```diff
@@ -117,6 +117,128 @@ def op_matrix(elements):
 I4_array[-4, -4] = 0
 I4 = op_matrix(I4_array)
 
+QNPI4_array = np.zeros((16, 16))
```
Is there a simpler way to implement the torch version?

obliviateandsurrender replied:
I had written the matrices as it is for readability and looking at the previous implementation. I've now used `np.diag` (for diagonal matrices) and `np.fliplr` ( for off-diagonal matrix), thus making it a bit simpler.

soranjh commented on pennylane/devices/torch_ops.py:
```diff
@@ -465,3 +587,27 @@ def DoubleExcitationMinus(phi, device):
     return e * I4(device) + c * UCos(device) + s * USin(device)
+
+
+def QuantumNumberPreservingOR(theta, device):
```
\phi?

obliviateandsurrender replied:
Fixed.

soranjh commented on pennylane/ops/qubit/qchem_ops.py:
```diff
@@ -464,3 +464,146 @@ def _matrix(cls, *params):
+
+    The overall matrix representation for this gate is:
+
+    .. math:: U(\phi) = \begin{bmatrix}
```
Some of the signs are different than those in the paper, please see my comment above.

soranjh commented on pennylane/ops/qubit/qchem_ops.py:
```diff
@@ -464,3 +464,146 @@ def _matrix(cls, *params):
+        varphi (float): rotation angle :math:`\varphi`
+        wires (Sequence[int]): the wires the operation acts on
+
+    **Example**
```
Will be good to give the output of the example.

soranjh commented on pennylane/ops/qubit/qchem_ops.py:
```diff
@@ -464,3 +464,146 @@ def _matrix(cls, *params):
+        c = math.cos(theta / 2)
+        s = math.sin(theta / 2)
+
+        return np.array(
```
Please see my comments regarding the signs.

obliviateandsurrender replied:
Done.

soranjh commented on pennylane/ops/qubit/qchem_ops.py:
```diff
@@ -464,3 +464,146 @@ def _matrix(cls, *params):
+
+    @staticmethod
+    def decomposition(theta, wires):
+        # This decomposition is the "upside down" version of that on p18 of https://arxiv.org/abs/2104.05695
```
Do they have the opposite convention for naming the wires?

obliviateandsurrender replied:
Yes. For the very same reason, their matrix for Given's rotation is flipped as well. In fact, if you look at [implementation](https://pennylane.readthedocs.io/en/stable/_modules/pennylane/ops/qubit/qchem_ops.html#DoubleExcitation) of `DoubleExcitation` operation, it uses the flipped version as well.

soranjh replied:
Good to mention somewhere that the sign flip is related to the opposite convention for naming the wires.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+from pennylane.operation import Operation, AnyWires
+
+
+class QuantumNumberPreservingU2(Operation):
```
Can we just use GateFabric as the name of the template?

obliviateandsurrender replied:
Yes. I have changed the name now.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+
+class QuantumNumberPreservingU2(Operation):
+    r"""Implements a local, expressive, and quantum-number-preserving ansatz using VQE circuit fabrics
+    proposed by Anselmetti *et al.* in `arXiv:2104.05692 <https://arxiv.org/abs/2104.05692>`_.
```
The correct link is: https://arxiv.org/abs/2104.05695

obliviateandsurrender replied:
Fixed.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+    r"""Implements a local, expressive, and quantum-number-preserving ansatz using VQE circuit fabrics
+    proposed by Anselmetti *et al.* in `arXiv:2104.05692 <https://arxiv.org/abs/2104.05692>`_.
+
+    This template prepares the :math:`2M` qubits trial state, where `M` is the number of spatial orbitals.
```
Is there any advantage in working with the number of spatial orbitals? Can we just use the number of spin-orbitals?

obliviateandsurrender replied:
I don't think there's any advantage as such. We are now using number of spin-orbitals.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+    proposed by Anselmetti *et al.* in `arXiv:2104.05692 <https://arxiv.org/abs/2104.05692>`_.
+
+    This template prepares the :math:`2M` qubits trial state, where `M` is the number of spatial orbitals.
+    It uses 4-local-nearest-neighbor-tessellation of alternating even and off spatial-orbitals-pair 2-parameter,
```
even and odd? I find this sentence a bit hard to read.

obliviateandsurrender replied:
I have rewritten the entire thing to make it sound crisper.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+    gate. In the :math:`\hat{Q}` gate we also allow inclusion of an optional constant :math:`\hat{\Pi}` gate,
+    with choices :math:`\hat{\Pi} \in \{\hat{I}, \text{QuantumNumberPreserving}_{OR}(\pi)\}`.
+
+    The circuit implementing the gate fabric layer for `M = 4` is shown below:
```
The `Q` should be `\hat{Q}`. You can also use `\hat{Q}(\phi, \theta)` for consistency.

obliviateandsurrender replied:
Fixed. In the figures as well.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+
+    |
+
+    The 2-parameter, 4-qubit :math:`\hat{Q}` gate is decomposed as follows:
```
The `Q` should be `\hat{Q}`.

obliviateandsurrender replied:
Done.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+        :width: 100%
+        :target: javascript:void(0);
+
+    |
```
It will be very informative to have a figure for the `DoubleExcitation` and `QuantumNumberPreservingOR` gates similar to those in Fig 2 of Anselmetti `et al.`

obliviateandsurrender replied:
I have now added the images for both `DoubleExcitation` and `OrbitalRotation`.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+    |
+
+    Args:
+        weights (tensor_like): Array of weights of shape ``(L, D, 2)``\.
+            ``L`` is the number of gate fabric layers and :math:`D=M-1`
+            is the number of :math:`\hat{Q}(\varphi, \theta)` gates per layer.
```
The parameters render with slightly different formats when you build the docs.

obliviateandsurrender replied:
Yes. I think it will render similar to this [page](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.templates.layers.ParticleConservingU2.html). Do I need to update this?

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+        wires (Iterable): wires that the template acts on
+        init_state (tensor_like): iterable or shape ``(len(wires),)`` tensor representing the Hartree-Fock state
+            used to initialize the wires
+        pi_gate_include (boolean): If `True`, sets the optional constant :math:`\Pi` to :math:`\text{QuantumNumberPreserving}_{OR}(\pi)`.
```
Maybe `include_pi = True` looks better?

obliviateandsurrender replied:
Done.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+           spin orbitals included in the active space.
+
+        #. The number of trainable parameters scales linearly with the number of layers as
+           :math:`2D(N-1)`.
```
The number of layers was named `L` before. I think this equation should be modified.

obliviateandsurrender replied:
Fixed.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+    par_domain = "A"
+
+    def __init__(
+        self, weights, wires=None, init_state=None, pi_gate_include=False, do_queue=True, id=None
```
Why there are `None` defaults for  `wires` and `init_state`?

obliviateandsurrender replied:
That was a mistake. These are fixed now.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+            raise ValueError(f"Inital state should be provided; got {init_state}")
+        self.init_state = qml.math.toarray(init_state)
+
+        self.pi_gate = pi_gate_include
```
Can we have one variable?

obliviateandsurrender replied:
I don't think we can. `self.init_state` determined the initial state of the wire before any layer of `GateFabric` is applied. Whereas, the `pi_gate` determines the initial `Pi` state for each of `Q` gates in the layer.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+
+        with qml.tape.QuantumTape() as tape:
+
+            qml.templates.BasisEmbedding(self.init_state, wires=self.wires)
```
You can use `BasisState` instead.

obliviateandsurrender replied:
I had tried using `BasisState` but it lead to the failure of the `test_adjoint` test case, quoting Basis state cannot be used after other gates. So, I decided to stick with `qml.templates.BasisEmbedding`.

soranjh commented on tests/gate_data.py:
```diff
@@ -438,3 +438,37 @@ def DoubleExcitationMinus(phi):
     return U
+
+
+def QuantumNumberPreservingOR(theta):
```
The name of the parameter for this gate is different than the one used in the figure.

obliviateandsurrender replied:
Fixed.

soranjh commented on tests/gate_data.py:
```diff
@@ -438,3 +438,37 @@ def DoubleExcitationMinus(phi):
+
+    return np.array(
+        [
+            [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
```
Some of the signs are different than those in the paper, please see my comment above.

obliviateandsurrender replied:
Done.

soranjh commented on pennylane/templates/layers/quantum_number_preserving_u2.py:
```diff
@@ -0,0 +1,216 @@
+                    if self.pi_gate:
+                        qml.QuantumNumberPreservingOR(np.pi, wires=wires)
+
+                    qml.DoubleExcitation(weight[layer][idx][0], wires=wires)
```
This gate replaces the QNP_PX gate in the original paper, right? If so, then it should be documented  in details in the docstring.

obliviateandsurrender replied:
The `DoubleExcitation` gate doesn't replace `QNP_PX`, but it actually is the same gate. If you look at the implementation [here](https://pennylane.readthedocs.io/en/stable/_modules/pennylane/ops/qubit/qchem_ops.html#DoubleExcitation), it uses the decomposition provided in the same paper. I have now mentioned in the docstring that we implement the diagonal-pair exchange gate, i.e., `QNP_PX` using `DoubleExcitation` gate.

soranjh reviewed (changes requested):
Thanks @obliviateandsurrender! Looking good. Left some comments and questions, Please also improve the docstring of the template by:
1. Explain why the QNP_PX is replace with DoubleExcitation.
2. Make sure that the variable names are consistent between the figures, the unctions and the docstring. 
3. Polish the text a bit more.

obliviateandsurrender commented:
Hello! @soranjh, thanks a lot for your comments!

I have incorporated the suggested changes and fixed the issues raised by you. Additionally, I have added new figures with consistent parameter names as requested. I have rewritten the docstring for both the template and `OrbitalRotation`. In the former, I have specified that we use `DoubleExcitation` for implementing the diagonal-pair exchange gate. However, I am now sure about the "specific" reason to mention because they are essentially the same gates (only with different names).

soranjh commented on pennylane/templates/layers/gate_fabric.py:
```diff
@@ -0,0 +1,235 @@
+
+    |
+
+    The individual :math:`\text{DoubleExcitation}(\theta)` and  :math:`\text{OrbitalRotation}(\varphi)` gates are decomposed using
+    Givens rotation gate :math:`G(\phi)`, which is implemented in pennylane with :math:`\text{SingleExcitation}(phi)`.
+
+    |
+
+    .. figure:: ../../_static/templates/layers/double_excitation_decomposition.png
+        :align: center
+        :width: 80%
+        :target: javascript:void(0);
+
+    .. figure:: ../../_static/templates/layers/orbital_rotation_decomposition.png
+        :align: center
+        :width: 80%
+        :target: javascript:void(0);
+
+    .. figure:: ../../_static/templates/layers/givens_rotation_decomposition.png
+        :align: center
+        :width: 100%
+        :target: javascript:void(0);
+
+    |
```
Actually, now that I look at the other templates, I think it is better to not have the decomposed gates here. They can appear in their own page. Maybe just mention specifically that the QNP_PX gate in the original ref is equivalent to Double excitation (are we sure that the CNOT gates in QNP_PX can be neglected?).

obliviateandsurrender replied:
I have removed the decompositions, and instead, just mentioned that the gates `QNP_OR` and `QNP_PX` given in the original references are equivalent to the `OrbitalRotation` and `DoubleExcitation` gates.

soranjh commented on pennylane/templates/layers/gate_fabric.py:
```diff
@@ -0,0 +1,235 @@
+
+    Args:
+        weights (tensor_like): Array of weights of shape ``(L, D, 2)``\,
+            where ``L`` is the number of gate fabric layers and ``D = N/2-1``\
```
```suggestion
            where ``L`` is the number of gate fabric layers and ``D = N/2-1`` \
```

obliviateandsurrender replied:
Fixed.

soranjh commented on pennylane/templates/layers/gate_fabric.py:
```diff
@@ -0,0 +1,235 @@
+
+    |
+
+    Args:
```
Please make sure that the format is consistent with the documentation [guide](https://pennylane.readthedocs.io/en/stable/development/guide/documentation.html#functions-and-methods). For instance, using capital vs small letter and using "." need to be consistent with the guideline.

obliviateandsurrender replied:
Yes, I have been following the guide and have been referring to other template documents too.

soranjh commented on pennylane/templates/layers/gate_fabric.py:
```diff
@@ -0,0 +1,235 @@
+            where ``L`` is the number of gate fabric layers and ``D = N/2-1``\
+            is the number of :math:`\hat{Q}(\varphi, \theta)` gates per layer.
+        wires (Iterable): wires that the template acts on
+        init_state (tensor_like): iterable or shape ``(len(wires),)`` tensor representing the Hartree-Fock state in Jordan-Wigner basis,
```
"of" shape? Also, the comma is needed in `(len(wires),)`?

obliviateandsurrender replied:
The shape of an array is always a tuple and the comma in `(len(wires),)` indicates that. It can also be seen in the `.shape` property of the Hartree Fock state array returned from `qml.qchem.hf_state`.

soranjh commented on pennylane/templates/layers/gate_fabric.py:
```diff
@@ -0,0 +1,235 @@
+    This template prepares the :math:`N` qubits trial state by applying :math:`L` layers of the gate fabric, which is composed
+    of 2-parameter 4-qubit gate elements :math:`\hat{Q}(\theta, \varphi)`. Each of the :math:`\hat{Q}(\theta, \varphi)` gate is itself
+    composed of two 1-parameter 4-qubit gates. The first gate is a spin-adapted spatial orbital rotation gate, which is implemented by
+    :math:`\text{OrbitalRotation}(\varphi)` gate. Whereas the second gate is a diagonal pair-exchange gate, which is implemented by the
```
Please use cross referencing for `OrbitalRotation` and `DoubleExcitation` as explained [here](https://pennylane.readthedocs.io/en/stable/development/guide/documentation.html#guiding-principles).

obliviateandsurrender replied:
Done.

soranjh commented on pennylane/ops/qubit/qchem_ops.py:
```diff
@@ -464,3 +464,160 @@ def _matrix(cls, *params):
+
+
+class OrbitalRotation(Operation):
+    r"""OrbitalRotation(varphi, wires)
```
Could not be just `phi`?

obliviateandsurrender replied:
Yes, it can be. I have now have changed the  `phi` in both the codebase and relevant images.

obliviateandsurrender commented:
After discussing this with @ixfoduap, it seems that it would be much better that I break this PR into two subsequent PRs - one for `OrbitalRotation` operation and another for the `GateFabric` template. Thus, I am closing this for now.
